src/UserSpaceInstrumentation/Attach.cpp

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "UserSpaceInstrumentation/Attach.h"
+
+#include <errno.h>
+#include <sys/ptrace.h>
+#include <sys/wait.h>
+
+#include <chrono>
+#include <memory>
+#include <set>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include "OrbitBase/GetProcessIds.h"
+#include "OrbitBase/Logging.h"
+#include "absl/strings/str_format.h"
+
+namespace orbit_user_space_instrumentation {
+
+using orbit_base::GetTidsOfProcess;
+
+namespace {
+
+bool ThreadExistsInProcess(pid_t pid, pid_t tid) {
+  const auto process_tids = orbit_base::GetTidsOfProcess(pid);
+  return std::find(process_tids.begin(), process_tids.end(), tid) != process_tids.end();
+}
+
+// Attaches to the thread using ptrace.
+// Returns true if the thread was halted and false if the thread did not exist anymore.
+[[nodiscard]] ErrorMessageOr<bool> AttachAndStopThread(pid_t pid, pid_t tid) {
+  if (ptrace(PTRACE_ATTACH, tid, 0, 0) < 0) {
+    if (ThreadExistsInProcess(pid, tid)) {
```

Comment:
probably better to check errno here

Reply:
(likely check for ESRCH) call to ThreadExistsInProcess in this case is flaky and could yield invalid result (false positive).

Reply:
Changed to errno. Sometimes errno is ESRCH. Weirdly errno is EPERM occationally.  Added a comment. This is only understandable after reading the kernel sources :/